### PR TITLE
ESC: deprecate speed(percent), introduce throttle(us).

### PR DESCRIPTION
**Pull request by rwaldron**
- Eliminates "history" memory leak
- Deprecates speed()
- Introduces throttle(us)

### CODE REVIEW

File: lib/esc.js [hunk 1]
```diff
@@ -209,34 +207,22 @@ function ESC(opts) {
...
     }
-
-    this.startAt = this.neutral;
   }
 
-  // Match either null or undefined, but not 0
-  if (this.startAt !== null && this.startAt !== undefined) {
-    this.speed(this.startAt);
-  }
+  this.throttle(this.neutral);
```
Comment (rwaldron) [inline lib/esc.js]: This is a safety precaution. No program should ever just "start" an ESC without explicit instruction. This is the safest thing I can think of.

File: lib/esc.js [hunk 2]
```diff
@@ -249,11 +235,11 @@ util.inherits(ESC, Emitter);
  * @param  {Float} speed 0...100 (full range)
  *
  * @return {ESC} instance
+ * @deprecated Will be deleted in version 1.0.0. Use throttle(us) instead.
  */
-
-ESC.prototype.speed = function(speed) {
+/* istanbul ignore next */
+ESC.prototype.speed = util.deprecate(function(speed) {
```
Comment (rwaldron) [inline lib/esc.js]: This whole function is trash. It did async work to speed up/down, which was really just papering over the fact that integers between 0-100 scales to ~18us for each step and that's awful, so let's get rid of this trash.

File: lib/esc.js [hunk 3]
```diff
@@ -333,24 +313,29 @@ ESC.prototype.speed = function(speed) {
...
+ * Throttle the ESC's speed by setting the pulse
+ *
+ * @param  {Integer} throttle pwmRange[0]...pwmRange[1] (full usec range)
+ *
+ * @return {ESC} instance
+ */
+ESC.prototype.throttle = function(pulse) {
+  this.write(this.pin, Fn.constrain(pulse, this.pwmRange[0], this.pwmRange[1]));
+  return this;
+};
```
Comment (rwaldron) [inline lib/esc.js]: This is what it should've always been.

File: lib/esc.js [hunk 4]
```diff
@@ -394,17 +379,7 @@ ESC.prototype.brake = function() {
...
-  var speed = this.type === "bidirectional" ? this.neutral : 0;
-
-  this.write(this.pin, Fn.fscale(speed, 0, 100, this.pwmRange[0], this.pwmRange[1]));
-
-  history.push({
-    timestamp: Date.now(),
-    speed: speed
-  });
-
+  this.write(this.pin, this.neutral);
```
Comment (rwaldron) [inline lib/esc.js]: `stop` and `brake` were basically the same thing... Let's figure out why

File: lib/esc.js (hunk 3 above)
Comment (dtex) [reply lib/esc.js]: Are you sure about 544-2400 being the ideal range for ESC's? That range comes from the Arduino servo library. I understand some ESC's will automatically adjust to fit the range they see, but I don't know what typical ESC's expect.

Comment (rwaldron) [reply lib/esc.js]: > Are you sure about 544-2400

As a default? Sure. User code can and should use the appropriate range for whatever platform and hardware is in use.

Comment (rwaldron) [reply lib/esc.js]: Upon further investigation, I believe that this range is _not_ appropriate for ESCs and will update this patch accordingly.

File: lib/esc.js (hunk 4 above)
Comment (dtex) [reply lib/esc.js]: I have a guess why it's like this. Some ESC's have the ability to brake and some don't (that's speculation, not fact). The default don't so they just replicate the behavior of ```stop```. That can be overridden in controller objects that do support it.

**Comment by HipsterBrown** [comment]
Finally tested this while preparing for the Nodeboats workshop at the upcoming JS Conf US. 

I ran the following code to test this patch:

```js
const five = require('johnny-five');
const keypress = require('keypress');

const board = new five.Board({
  port: '/dev/tty.blobfish-DevB',
});

board.on('error', error => {
  console.error(error);
  process.exit(1);
});

board.on('ready', () => {
  let speed = 50;
  const led = new five.Led(13);
  const esc = new five.ESC({
    device: 'FORWARD_REVERSE',
    neutral: speed,
    pin: 11,
  });

  // just to make sure the program is running
  led.blink(500);

  function controller(_, key) {
    let isThrottle = false;

    if (key && key.shift) {
      if (key.name === 'up' && speed <= 100) {
        speed += 1;
        isThrottle = true;
      }

      if (key.name === 'down' && speed >= 0) {
        speed -= 1;
        isThrottle = true;
      }

      if (isThrottle) {
        console.log('Setting speed: ', speed);

        esc.throttle(
          five.Fn.scale(speed, 0, 100, esc.pwmRange[0], esc.pwmRange[1]),
        );
      }
    }
  }

  keypress(process.stdin);

  process.stdin.on('keypress', controller);
  process.stdin.setRawMode(true);
  process.stdin.resume();
});
```

It ran perfectly. 👌

I noticed a comment about updating the default PWM range (https://github.com/rwaldron/johnny-five/pull/1484#discussion_r204090707) but didn't notice a change in the PR. So that still may be pending.

**Comment by dtex** [comment]
The light finally went on re: why throttle is a better method name than speed, but I think it's weird that Johnny-Five doesn't abstract away the need to pass values in microseconds. It seems like that would be in keeping with other classes in J5. 

There was a plan to migrate input values like this to a [0.0 - 1.0] or [-1.0, 1.0] range. Why not go ahead and do that here?

**Comment by HipsterBrown** [comment]
So throttle or some other method would just accept a range? Or you would be able to set the range on the ESC class instantiation and use a value in that range when calling throttle()?

**Comment by dtex** [comment]
No, it was a single set range for all output devices whose values do not correspond to some real-world measure. I misremembered the details though, it's not 0-1.0, it's 0%-100%.

Here's the issue: https://github.com/rwaldron/johnny-five/issues/370

**Comment by rwaldron** [comment]
@dtex your intuition is right, but I think I landed on this for a few reasons: 

- `throttle()` is meant to mimic the trigger or thumb stick on an actual RC transmitter. Those physically default to whatever is "neutral" (or throttling no speed at all). 
- "throttling" _might_ be bidirectional, so what does 0% mean?
- Mapping 0-100% to μs would cause a massive loss in precision, noted in my comment above:
    > It did async work to speed up/down, which was really just papering over the fact that integers between 0-100 scales to ~18us for each step and that's awful, so let's get rid of this trash.

    That meant that motors attached to ESCs might jump 18-19μs when throttling up 1%, which is physically dangerous (as @buritica and I experienced first hand about 5 years ago when we tried to build a hovering "jellyfish" drone)

So it's not as "nice", but it's safe and accurate.

**Comment by dtex** [comment]
> Mapping 0-100% to μs would cause a massive loss in precision

I didn't mean to suggest we were limited to integers. I should have written 0.000% - 100.000%. Values input in that range will be mapped to the nearest whole microsecond value before being sent to the IO plugin.

> That meant that motors attached to ESCs might jump 18-19μs when throttling up 1%

I'm on board with that. It's the same underlying principle for why J5 servo writes are in microseconds now. Servos can be positioned in fractions of degrees.

**Comment by rwaldron** [comment]
@dtex 

This morning I was still wondering about this: 

> "throttling" might be bidirectional, so what does 0% mean?

... but I think I've convinced myself that it's safe to allow both: 

- `throttle(0-100%)`
- `throttle(1000-2000usec)`

**Comment by buritica** [comment]
I remember very well

File: eg/esc-blobfish.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      }
+    }
+  }
+
+  keypress(process.stdin);
+
+  process.stdin.on("keypress", controller);
+  process.stdin.setRawMode(true);
+  process.stdin.resume();
+});
```
Comment (rwaldron) [inline eg/esc-blobfish.js]: @HipsterBrown ^^^ I've been using this to test all day... it feels really good

**Comment by rwaldron** [comment]
I've made a number of changes that meet the need set out by @dtex, achieve my goals of simplifying and are ready for @HipsterBrown to use in the NodeBoats workshop. I want to merge this, but I'd like @HipsterBrown to verify that his boat works properly

**Comment by HipsterBrown** [comment]
This works for my nodeboat setup. 👍 LGTM

**Comment by dtex** [comment]
LGTM!

I think someday ESC and Motor should be refactored to have the same surface API. It's weird to me that they are different.

**Comment by rwaldron** [comment]
> I think someday ESC and Motor should be refactored to have the same surface API. It's weird to me that they are different.

I look forward to this potentially contentious debate <3

**Pull request merged**